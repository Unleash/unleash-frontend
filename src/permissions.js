export const ADMIN = 'ADMIN';
export const CREATE_FEATURE = 'CREATE_FEATURE';
export const UPDATE_FEATURE = 'UPDATE_FEATURE';
export const DELETE_FEATURE = 'DELETE_FEATURE';
export const CREATE_STRATEGY = 'CREATE_STRATEGY';
export const UPDATE_STRATEGY = 'UPDATE_STRATEGY';
export const DELETE_STRATEGY = 'DELETE_STRATEGY';
export const UPDATE_APPLICATION = 'UPDATE_APPLICATION';
export const CREATE_CONTEXT_FIELD = 'CREATE_CONTEXT_FIELD';
export const UPDATE_CONTEXT_FIELD = 'UPDATE_CONTEXT_FIELD';
export const DELETE_CONTEXT_FIELD = 'DELETE_CONTEXT_FIELD';

export function hasPermission(user, permission) {
    return (
        user &&
        (!user.permissions || user.permissions.indexOf(ADMIN) !== -1 || user.permissions.indexOf(permission) !== -1)
    );
}
