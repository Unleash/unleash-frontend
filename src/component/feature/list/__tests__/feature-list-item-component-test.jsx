import React from 'react';
import { MemoryRouter } from 'react-router-dom';

import Feature from '../list-item-component';
import renderer from 'react-test-renderer';
import { UPDATE_FEATURE } from '../../../../permissions';

jest.mock('../feature-type-container');

test('renders correctly with one feature', () => {
    const feature = {
        name: 'Another',
        description: "another's description",
        enabled: false,
        stale: false,
        strategies: [
            {
                name: 'gradualRolloutRandom',
                parameters: {
                    percentage: 50,
                },
            },
        ],
        createdAt: '2018-02-04T20:27:52.127Z',
    };
    const featureMetrics = { lastHour: {}, lastMinute: {}, seenApps: {} };
    const settings = { sort: 'name' };
    const tree = renderer.create(
        <MemoryRouter>
            <Feature
                key={0}
                settings={settings}
                metricsLastHour={featureMetrics.lastHour[feature.name]}
                metricsLastMinute={featureMetrics.lastMinute[feature.name]}
                feature={feature}
                toggleFeature={jest.fn()}
                hasPermission={permission => permission === UPDATE_FEATURE}
            />
        </MemoryRouter>
    );

    expect(tree).toMatchSnapshot();
});

test('renders correctly with one feature without permission', () => {
    const feature = {
        name: 'Another',
        description: "another's description",
        enabled: false,
        stale: false,
        strategies: [
            {
                name: 'gradualRolloutRandom',
                parameters: {
                    percentage: 50,
                },
            },
        ],
        createdAt: '2018-02-04T20:27:52.127Z',
    };
    const featureMetrics = { lastHour: {}, lastMinute: {}, seenApps: {} };
    const settings = { sort: 'name' };
    const tree = renderer.create(
        <MemoryRouter>
            <Feature
                key={0}
                settings={settings}
                metricsLastHour={featureMetrics.lastHour[feature.name]}
                metricsLastMinute={featureMetrics.lastMinute[feature.name]}
                feature={feature}
                toggleFeature={jest.fn()}
                hasPermission={() => false}
            />
        </MemoryRouter>
    );

    expect(tree).toMatchSnapshot();
});
