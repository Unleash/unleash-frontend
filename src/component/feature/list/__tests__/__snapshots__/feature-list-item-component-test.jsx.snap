// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with one feature 1`] = `
<li
  className="MuiListItem-root MuiListItem-gutters"
  disabled={false}
>
  <span
    className="listItemMetric"
  >
    <svg
      className="mdl-color-text--grey-300"
      viewBox="0 0 24 24"
    >
      <path
        d="M17.3,18C19,16.5 20,14.4 20,12A8,8 0 0,0 12,4A8,8 0 0,0 4,12C4,14.4 5,16.5 6.7,18C8.2,16.7 10,16 12,16C14,16 15.9,16.7 17.3,18M12,2A10,10 0 0,1 22,12A10,10 0 0,1 12,22A10,10 0 0,1 2,12A10,10 0 0,1 12,2M7,9A1,1 0 0,1 8,10A1,1 0 0,1 7,11A1,1 0 0,1 6,10A1,1 0 0,1 7,9M10,6A1,1 0 0,1 11,7A1,1 0 0,1 10,8A1,1 0 0,1 9,7A1,1 0 0,1 10,6M17,9A1,1 0 0,1 18,10A1,1 0 0,1 17,11A1,1 0 0,1 16,10A1,1 0 0,1 17,9M14.4,6.1C14.9,6.3 15.1,6.9 15,7.4L13.6,10.8C13.8,11.1 14,11.5 14,12A2,2 0 0,1 12,14A2,2 0 0,1 10,12C10,11 10.7,10.1 11.7,10L13.1,6.7C13.3,6.1 13.9,5.9 14.4,6.1Z"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    className="listItemToggle"
  >
    <span
      className="MuiSwitch-root"
    >
      <span
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiSwitch-switchBase MuiSwitch-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
        title="Toggle Another"
      >
        <span
          className="MuiIconButton-label"
        >
          <input
            checked={false}
            className="PrivateSwitchBase-input-4 MuiSwitch-input"
            disabled={false}
            onChange={[Function]}
            type="checkbox"
          />
          <span
            className="MuiSwitch-thumb"
          />
        </span>
      </span>
      <span
        className="MuiSwitch-track"
      />
    </span>
  </span>
  <span
    className="listItemLink"
  >
    <a
      className="listLink truncate"
      href="/features/strategies/Another"
      onClick={[Function]}
    >
      <span
        className="toggleName"
      >
        Another
         
      </span>
      <small>
        <time
          dateTime="2018-02-04T20:27:52.127Z"
          title="2018-02-04T20:27:52.127Z"
        >
          3 years ago
        </time>
      </small>
      <div>
        <span
          className="truncate"
        >
          another's description
        </span>
      </div>
    </a>
  </span>
  <span
    className="listItemStrategies hideLt920"
  />
  <span />
</li>
`;

exports[`renders correctly with one feature without permission 1`] = `
<li
  className="MuiListItem-root MuiListItem-gutters"
  disabled={false}
>
  <span
    className="listItemMetric"
  >
    <svg
      className="mdl-color-text--grey-300"
      viewBox="0 0 24 24"
    >
      <path
        d="M17.3,18C19,16.5 20,14.4 20,12A8,8 0 0,0 12,4A8,8 0 0,0 4,12C4,14.4 5,16.5 6.7,18C8.2,16.7 10,16 12,16C14,16 15.9,16.7 17.3,18M12,2A10,10 0 0,1 22,12A10,10 0 0,1 12,22A10,10 0 0,1 2,12A10,10 0 0,1 12,2M7,9A1,1 0 0,1 8,10A1,1 0 0,1 7,11A1,1 0 0,1 6,10A1,1 0 0,1 7,9M10,6A1,1 0 0,1 11,7A1,1 0 0,1 10,8A1,1 0 0,1 9,7A1,1 0 0,1 10,6M17,9A1,1 0 0,1 18,10A1,1 0 0,1 17,11A1,1 0 0,1 16,10A1,1 0 0,1 17,9M14.4,6.1C14.9,6.3 15.1,6.9 15,7.4L13.6,10.8C13.8,11.1 14,11.5 14,12A2,2 0 0,1 12,14A2,2 0 0,1 10,12C10,11 10.7,10.1 11.7,10L13.1,6.7C13.3,6.1 13.9,5.9 14.4,6.1Z"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    className="listItemToggle"
  >
    <span
      className="MuiSwitch-root"
    >
      <span
        aria-disabled={true}
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-disabled-3 Mui-disabled Mui-disabled Mui-disabled"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        title="Toggle Another"
      >
        <span
          className="MuiIconButton-label"
        >
          <input
            checked={false}
            className="PrivateSwitchBase-input-4 MuiSwitch-input"
            disabled={true}
            onChange={[Function]}
            type="checkbox"
          />
          <span
            className="MuiSwitch-thumb"
          />
        </span>
      </span>
      <span
        className="MuiSwitch-track"
      />
    </span>
  </span>
  <span
    className="listItemLink"
  >
    <a
      className="listLink truncate"
      href="/features/strategies/Another"
      onClick={[Function]}
    >
      <span
        className="toggleName"
      >
        Another
         
      </span>
      <small>
        <time
          dateTime="2018-02-04T20:27:52.127Z"
          title="2018-02-04T20:27:52.127Z"
        >
          3 years ago
        </time>
      </small>
      <div>
        <span
          className="truncate"
        >
          another's description
        </span>
      </div>
    </a>
  </span>
  <span
    className="listItemStrategies hideLt920"
  />
  <span />
</li>
`;
